src/games/rcll/workpieces.clp

```diff
@@ -0,0 +1,373 @@
+    Update workpiece by read event information (mps-status-feedback)"
+    (workpiece-tracking (enabled TRUE))
+    (gamestate (phase PRODUCTION) (game-time ?gt))
+    ?mf <- (mps-status-feedback ?m-name BARCODE ?id)
+    (machine (name ?m-name) (state ~BROKEN) (team ?team))
+    =>
+    (retract ?mf)
+    (printout t "Workpiece " ?id ": at " ?m-name ", availabe!"crlf)
+    (if (any-factp ((?wp workpiece)) (eq ?wp:id ?id)) then
+       ;Update existing
+       (do-for-fact ((?workpiece workpiece)) (eq ?workpiece:id ?id)
+         (modify ?workpiece (at-machine ?m-name) (state AVAILABLE) (visible ?gt)))
+      else
+        ;Learn new
+        (assert (workpiece (at-machine ?m-name) (state AVAILABLE) (visible ?gt) (id ?id)
+                           (team ?team) (base-color (workpiece-base-color-by-id ?id))))
+    )
+)
+
+(defrule workpiece-update-not-in-production
+    "workpiece update received at wrong time"
+    (gamestate (phase ~PRODUCTION))
+    (workpiece-tracking (enabled TRUE))
+    ?mf <- (mps-status-feedback ?machine BARCODE ?id)
+	=>
+    (retract ?mf)
+    (printout warn "Received workpiece update for " ?id " while not in production" crlf)
+)
+
+(defrule workpiece-update-machine-broken
+    "workpiece update received at while machine is broken"
+    (gamestate (phase PRODUCTION))
+    (workpiece-tracking (enabled TRUE))
+    ?mf <- (mps-status-feedback ?machine BARCODE ?id)
+    (machine (name ?machine) (state BROKEN))
+	=>
+    (retract ?mf)
+    (printout warn "Received workpiece update for " ?id " while " ?machine "  BROKEN" crlf)
+)
+
+(defrule workpiece-update-tracking-disabled
+    "workpiece update received but tracking is disabled"
+    (gamestate (phase PRODUCTION))
+    (workpiece-tracking (enabled FALSE))
+    ?mf <- (mps-status-feedback ?machine BARCODE ?id)
+	=>
+    (retract ?mf)
+    (printout warn "Received workpiece update for " ?id " but workpiece tracking is disabled" crlf)
+)
+
+(defrule workpiece-update-retrieved
+    "Workpiece no longer at machine"
+    (workpiece-tracking (enabled TRUE))
+    (machine (name ?m-name) (state WAIT-IDLE|BROKEN))
+    (workpiece (id ?id) (at-machine ?m-name) (state AVAILABLE))
+    =>
+    (do-for-all-facts ((?workpiece workpiece)) (and (eq ?workpiece:at-machine ?m-name)
+                                               (eq ?workpiece:state AVAILABLE))
+      (printout t "Workpiece " ?workpiece:id ": at " ?m-name ", retrieved!" crlf)
+      (modify ?workpiece (state RETRIEVED))
+    )
+)
+
+;-----------------------------Order assignment---------------------------------
+(defrule workpiece-assign-order
+   "Assign order to workpiece.
+   This has a very specific assumption about the ring colors. We assume that
+   the color of the first ring fully determines the complexity class. Furthermore
+   we assume that there is only a single order of complexity >0 each. Hence, knowing
+   the first ring will immediately determine the order of the workpiece."
```

Comment:
That is a good Idea but!
As a matter of fact, this assumption doesn't even completely hold anymore. Which makes this documentation even more wrong, thanks for pointing it out! 
I have noticed 2 C1 identical orders. This assumption was a part of the old code I extended. To some extent I have adapted to not having this assumption anymore by having the following assumptions.
- The orders can be resigened and re-assigened to a workpiece at anypoint when its needed.
- Points calculated and given for a resigend order are recalculated when assigened a new.
- Workpieces are always verified against the confirmed delivery and rectifed if needed.
(please, keep an eye on those during testing)

Yet, A recent test case that your review had sparked makes me reconsider ensuring the "unique order prefix assumption". That is, related to the allowance of scoring on identical operations for the identical orders. atm. the ambiouty could lead us to prevent a team of scoring points inteded for an order identical to the one we are tracking.
Senario:
- Order 2 and 8 are identical C1 (b:BLACK|r1:BLUE|c:SILVER)
- a team produces 2 and 8 at the same time.
- we assume both workpieces to track order 8.
We will only award the mount ring blue once (since order 8 quantitiy is 1). If we deliver this tracking error should be rectified and the points intended for order 2 will be given. The issue will be if they did not delivery the order 2.

I need to think about if we want to ensure the "unique prefix assumption" (for anyother reasons) or simply adapt that case to not having it!.

what do you think?

Reply:
I took a closer look into this,
Violation of the assumption could lead us to many corner cases that tracking needs to make sure of actively correcting. The ambiguity generated is simply too annoying and sometimes there is actually no way to treat it till delivery is attempted by the team (which we already handle atm, our tracking error is only a problem when 2 orders with identical rings are not delivered)

The main issue was that we have 2 C1 orders posted, this violates the last part of the assumption
```
Furthermore, we assume that there is only a single order of complexity >0 each.
```